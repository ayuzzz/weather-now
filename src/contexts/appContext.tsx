"use client";

import React, { createContext, useContext, useState, ReactNode } from "react";
import { City } from "@/models/city";

interface AppContextType {
  theme: "light" | "dark";
  setTheme: (theme: "light" | "dark") => void;
  currentCity: City;
  setCurrentCity: (city: City) => void;
}

const AppContext = createContext<AppContextType | undefined>(undefined);

export const AppProvider = ({ children }: { children: ReactNode }) => {
  const [theme, setTheme] = useState<"light" | "dark">("light");
  const [currentCity, setCurrentCity] = useState<City>({
    id: 1277333,
    name: "Bengaluru",
    latitude: 12.97194,
    longitude: 77.59369,
    timezone: "Asia/Kolkata",
    country: "India",
  });

  return (
    <AppContext.Provider
      value={{ theme, setTheme, currentCity, setCurrentCity }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = (): AppContextType => {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error("useAppContext must be used within an AppProvider");
  }
  return context;
};
