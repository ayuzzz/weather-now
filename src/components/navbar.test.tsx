import { AppProvider } from "@/contexts/appContext";
import Navbar from "./navbar";
import { fireEvent, render, screen } from "@testing-library/react";

describe("Navbar", () => {
  it("renders correctly", () => {
    const { container } = render(
      <AppProvider>
        <Navbar />
      </AppProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("render the urls correctly", () => {
    render(
      <AppProvider>
        <Navbar />
      </AppProvider>
    );

    expect(screen.getByText("Dashboard")).toHaveAttribute("href", "/");
    expect(screen.getByText("Maps")).toHaveAttribute("href", "/maps");
    expect(screen.getByText("Settings")).toHaveAttribute("href", "/settings");
  });

  it("should have a logo", () => {
    const { container } = render(
      <AppProvider>
        <Navbar />
      </AppProvider>
    );
    const logo = container.querySelector("img");
    expect(logo).toHaveProperty("src", expect.stringContaining("favicon.png"));
    expect(logo).toHaveProperty("alt", "Logo");
    expect(logo).toBeInTheDocument();
  });
});

describe("Theme Toggle", () => {
  it("should render the theme toggle button", () => {
    render(
      <AppProvider>
        <Navbar />
      </AppProvider>
    );

    const button = screen.getByRole("button", { name: /theme toggle/i });
    expect(button).toBeInTheDocument();
  });

  it("should toggle the theme on click", () => {
    render(
      <AppProvider>
        <Navbar />
      </AppProvider>
    );

    const button = screen.getByRole("button", { name: /theme toggle/i });
    const icon = button.querySelector("svg");

    let iconPath = icon?.querySelector("path");
    expect(iconPath).toHaveAttribute(
      "d",
      "M9.528 1.718a.75.75 0 0 1 .162.819A8.97 8.97 0 0 0 9 6a9 9 0 0 0 9 9 8.97 8.97 0 0 0 3.463-.69.75.75 0 0 1 .981.98 10.503 10.503 0 0 1-9.694 6.46c-5.799 0-10.5-4.7-10.5-10.5 0-4.368 2.667-8.112 6.46-9.694a.75.75 0 0 1 .818.162Z"
    ); // HiMoon path data

    // Click to toggle theme
    fireEvent.click(button);

    // After clicking: Dark mode (should show sun icon)
    iconPath = button.querySelector("svg")?.querySelectorAll("path")[1];
    expect(iconPath).toHaveAttribute(
      "d",
      "M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58a.996.996 0 0 0-1.41 0 .996.996 0 0 0 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37a.996.996 0 0 0-1.41 0 .996.996 0 0 0 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0a.996.996 0 0 0 0-1.41l-1.06-1.06zm1.06-10.96a.996.996 0 0 0 0-1.41.996.996 0 0 0-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36a.996.996 0 0 0 0-1.41.996.996 0 0 0-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
    ); // MdLightMode path data
  });
});
