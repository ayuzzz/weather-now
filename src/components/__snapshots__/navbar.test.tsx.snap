// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<div>
  <div
    class="navbar"
  >
    <div
      class="logo"
    >
      <a
        href="/"
      >
        <img
          alt="Logo"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=%2Ffavicon.png&w=128&q=75"
          srcset="/_next/image?url=%2Ffavicon.png&w=64&q=75 1x, /_next/image?url=%2Ffavicon.png&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
      </a>
      <a
        class="logo_link"
        href="/"
      >
        <h1>
          Weather Now
        </h1>
      </a>
    </div>
    <div
      class="nav_items"
    >
      <div
        class="flex_center"
      >
        <svg
          class="nav_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.03 2.59a1.501 1.501 0 0 1 1.94 0l7.5 6.363a1.5 1.5 0 0 1 .53 1.144V19.5a1.5 1.5 0 0 1-1.5 1.5h-5.75a.75.75 0 0 1-.75-.75V14h-2v6.25a.75.75 0 0 1-.75.75H4.5A1.5 1.5 0 0 1 3 19.5v-9.403c0-.44.194-.859.53-1.144ZM12 3.734l-7.5 6.363V19.5h5v-6.25a.75.75 0 0 1 .75-.75h3.5a.75.75 0 0 1 .75.75v6.25h5v-9.403Z"
          />
        </svg>
        <a
          class="nav_item"
          href="/"
        >
          Dashboard
        </a>
      </div>
      <div
        class="flex_center"
      >
        <svg
          class="nav_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 32 32"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M 28 4.46875 L 26.59375 5.09375 L 19.96875 7.9375 L 12.34375 5.0625 L 11.96875 4.9375 L 11.59375 5.09375 L 4.59375 8.09375 L 4 8.34375 L 4 27.53125 L 5.40625 26.90625 L 12.03125 24.0625 L 19.65625 26.9375 L 20.03125 27.0625 L 20.40625 26.90625 L 27.40625 23.90625 L 28 23.65625 Z M 13 7.4375 L 19 9.6875 L 19 24.5625 L 13 22.3125 Z M 11 7.5 L 11 22.34375 L 6 24.5 L 6 9.65625 Z M 26 7.5 L 26 22.34375 L 21 24.5 L 21 9.65625 Z"
          />
        </svg>
        <a
          class="nav_item"
          href="/maps"
        >
          Maps
        </a>
      </div>
      <div
        class="flex_center"
      >
        <svg
          class="nav_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M19.43 12.98c.04-.32.07-.64.07-.98 0-.34-.03-.66-.07-.98l2.11-1.65c.19-.15.24-.42.12-.64l-2-3.46a.5.5 0 0 0-.61-.22l-2.49 1c-.52-.4-1.08-.73-1.69-.98l-.38-2.65A.488.488 0 0 0 14 2h-4c-.25 0-.46.18-.49.42l-.38 2.65c-.61.25-1.17.59-1.69.98l-2.49-1a.566.566 0 0 0-.18-.03c-.17 0-.34.09-.43.25l-2 3.46c-.13.22-.07.49.12.64l2.11 1.65c-.04.32-.07.65-.07.98 0 .33.03.66.07.98l-2.11 1.65c-.19.15-.24.42-.12.64l2 3.46a.5.5 0 0 0 .61.22l2.49-1c.52.4 1.08.73 1.69.98l.38 2.65c.03.24.24.42.49.42h4c.25 0 .46-.18.49-.42l.38-2.65c.61-.25 1.17-.59 1.69-.98l2.49 1c.06.02.12.03.18.03.17 0 .34-.09.43-.25l2-3.46c.12-.22.07-.49-.12-.64l-2.11-1.65zm-1.98-1.71c.04.31.05.52.05.73 0 .21-.02.43-.05.73l-.14 1.13.89.7 1.08.84-.7 1.21-1.27-.51-1.04-.42-.9.68c-.43.32-.84.56-1.25.73l-1.06.43-.16 1.13-.2 1.35h-1.4l-.19-1.35-.16-1.13-1.06-.43c-.43-.18-.83-.41-1.23-.71l-.91-.7-1.06.43-1.27.51-.7-1.21 1.08-.84.89-.7-.14-1.13c-.03-.31-.05-.54-.05-.74s.02-.43.05-.73l.14-1.13-.89-.7-1.08-.84.7-1.21 1.27.51 1.04.42.9-.68c.43-.32.84-.56 1.25-.73l1.06-.43.16-1.13.2-1.35h1.39l.19 1.35.16 1.13 1.06.43c.43.18.83.41 1.23.71l.91.7 1.06-.43 1.27-.51.7 1.21-1.07.85-.89.7.14 1.13zM12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 6c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
          />
        </svg>
        <a
          class="nav_item"
          href="/settings"
        >
          Settings
        </a>
      </div>
    </div>
    <div
      class="theme_toggle"
    >
      <button
        aria-label="Theme toggle"
        class="btn_theme_toggle flex_center"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.528 1.718a.75.75 0 0 1 .162.819A8.97 8.97 0 0 0 9 6a9 9 0 0 0 9 9 8.97 8.97 0 0 0 3.463-.69.75.75 0 0 1 .981.98 10.503 10.503 0 0 1-9.694 6.46c-5.799 0-10.5-4.7-10.5-10.5 0-4.368 2.667-8.112 6.46-9.694a.75.75 0 0 1 .818.162Z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
